index.ts

```diff
@@ -4,38 +4,69 @@ import { spawn } from 'child_process';
 
 setGracefulCleanup();
 
-type ProjectMutator = (project: Project) => void | Promise<void>;
+/**
+ * A callback used by some of classes to obtain a {@link Project}.
+ * 
+ * You are supposed to pass such a callback. In it, you should create and return a {@link Project}
+ * instance.
+ * 
+ * @returns A {@link Project} instance, optionally wrapped into a promise.
+ */
+
+export type CallbackCreateProject = () => Project | Promise<Project>;
+
+/**
+ * A callback used to customize an existing {@link Scenario}. In it, you receive a {@link Project}
+ * instance and you can do stuff with it in order to adjust the project's codebase to the specific
+ * needs of a given scenario.
+ * 
+ */
+export type CallbackMutateProject = (project: Project) => void | Promise<void>;
 
 export { Project };
 
-type State =
-  | {
-      type: 'root';
-      root: () => Project | Promise<Project>;
-    }
-  | {
-      type: 'derived';
-      parent: Scenarios;
-      variants: Record<string, ProjectMutator[]>;
-    };
```

Comment:
As we discussed on the office hours, this type is entirely private and self-contained so it should go back to being not exported.

If we add documentation to it, the most useful documentation would be on its individual fields. The object itself is really just an help to typescript so we know which fields are available in which states.

Reply:
@ef4 Done!